File: backend/src/agents/OrchestratorAgent.ts
import { BaseAgent } from './BaseAgent';
import { AgentResponse, AgentConfig } from './AgentInterface';
import { EventStream, DefaultEventStream } from './EventStream';
import { LLMProvider } from './LLMProvider';
import { Action } from './Action';
import { Observation } from './Observation';
import { AgentRegistry } from './AgentRegistry';
import { ChatOpenAI } from '@langchain/openai';
import { ChatGoogleGenerativeAI } from '@langchain/google-genai';
import { AgentExecutor, createOpenAIFunctionsAgent } from 'langchain/agents';
import { ChatPromptTemplate } from '@langchain/core/prompts';
import { StructuredTool } from '@langchain/core/tools';
import { BufferMemory } from 'langchain/memory';
import { PDFLoader } from 'langchain/document_loaders/fs/pdf';
import { TextLoader } from 'langchain/document_loaders/fs/text';
import { RecursiveCharacterTextSplitter } from 'langchain/text_splitter';
import { MemoryVectorStore } from 'langchain/vectorstores/memory';
import { OpenAIEmbeddings } from '@langchain/openai';
import { HuggingFaceInferenceEmbeddings } from '@langchain/community/embeddings/hf';

/**
 * Agent responsible for coordinating between other agents
 */
export class OrchestratorAgent extends BaseAgent {
  private agentExecutor: AgentExecutor | null = null;
  private vectorStores: Map<string, any> = new Map();
  private agentRegistry: AgentRegistry;
  
  /**
   * Create a new OrchestratorAgent
   * @param eventStream The event stream for the agent
   * @param llmProvider The LLM provider for the agent
   */
  constructor(eventStream: EventStream, llmProvider: LLMProvider) {
    super(eventStream, llmProvider);
    this.agentRegistry = AgentRegistry.getInstance();
  }
  
  /**
   * Initialize the agent
   * @param config Configuration for the agent
   */
  async initialize(config: AgentConfig): Promise<void> {
    await super.initialize(config);
    
    // Set default system message if not provided
    if (!this.systemMessage) {
      this.systemMessage = `You are an Orchestrator Agent that coordinates the work of specialized agents to complete complex tasks.
Your job is to:
1. Understand the user's request
2. Break it down into subtasks
3. Assign each subtask to the appropriate specialized agent
4. Combine the results into a coherent response

You have access to the following specialized agents:
- ThinkingAgent: Analyzes requests and creates execution plans
- DeveloperAgent: Generates code based on specifications
- EditorAgent: Reviews and improves code

Always think step by step and use the most appropriate agent for each subtask.`;
    }
  }
  
  /**
   * Create a chat model based on provider and API key
   */
  private createChatModel(provider: string, apiKey: string) {
    switch (provider.toLowerCase()) {
      case 'openai':
        return new ChatOpenAI({
          openAIApiKey: apiKey,
          modelName: 'gpt-3.5-turbo',
          temperature: 0.7,
        });
      case 'gemini':
        return new ChatGoogleGenerativeAI({
          apiKey: apiKey,
          modelName: 'gemini-pro',
          temperature: 0.7,
        });
      default:
        throw new Error(`Unsupported provider: ${provider}`);
    }
  }
  
  /**
   * Initialize the LangChain agent executor
   */
  private async initializeAgentExecutor(provider: string, apiKey: string): Promise<AgentExecutor> {
    if (this.agentExecutor) {
      return this.agentExecutor;
    }
    
    const model = this.createChatModel(provider, apiKey);
    const memory = new BufferMemory({
      memoryKey: 'chat_history',
      returnMessages: true,
    });
    
    // Create specialized agents if they don't exist
    const thinkingAgentId = 'thinking-agent';
    const developerAgentId = 'developer-agent';
    const editorAgentId = 'editor-agent';
    
    let thinkingAgent = this.agentRegistry.getAgent(thinkingAgentId);
    let developerAgent = this.agentRegistry.getAgent(developerAgentId);
    let editorAgent = this.agentRegistry.getAgent(editorAgentId);
    
    if (!thinkingAgent) {
      thinkingAgent = await this.agentRegistry.createAgent(
        'thinking',
        { id: thinkingAgentId, name: 'ThinkingAgent', description: 'Analyzes requests and creates execution plans' },
        new DefaultEventStream(),
        this.llmProvider
      );
    }
    
    if (!developerAgent) {
      developerAgent = await this.agentRegistry.createAgent(
        'developer',
        { id: developerAgentId, name: 'DeveloperAgent', description: 'Generates code based on specifications' },
        new DefaultEventStream(),
        this.llmProvider
      );
    }
    
    if (!editorAgent) {
      editorAgent = await this.agentRegistry.createAgent(
        'editor',
        { id: editorAgentId, name: 'EditorAgent', description: 'Reviews and improves code' },
        new DefaultEventStream(),
        this.llmProvider
      );
    }
    
    // Define tools for the orchestrator
    const tools = [
      new StructuredTool({
        name: 'analyze_request',
        description: 'Analyze the user request and break it down into steps',
        schema: {
          type: 'object',
          properties: {
            request: {
              type: 'string',
              description: 'The user request to analyze',
            },
          },
          required: ['request'],
        },
        func: async ({ request }) => {
          const observation = Observation.createUserMessageObservation('user', request);
          const action = await thinkingAgent.process(observation);
          return JSON.stringify(action.data);
        },
      }),
      new StructuredTool({
        name: 'generate_code',
        description: 'Generate code based on specifications',
        schema: {
          type: 'object',
          properties: {
            specifications: {
              type: 'string',
              description: 'The specifications for the code to generate',
            },
            language: {
              type: 'string',
              description: 'The programming language to use',
            },
            fileType: {
              type: 'string',
              description: 'The type of file to generate',
            },
          },
          required: ['specifications', 'language', 'fileType'],
        },
        func: async ({ specifications, language, fileType }) => {
          const observation = Observation.createUserMessageObservation('user', specifications);
          observation.data.context = { language, fileType };
          const action = await developerAgent.process(observation);
          return JSON.stringify(action.data);
        },
      }),
      new StructuredTool({
        name: 'review_code',
        description: 'Review and improve code',
        schema: {
          type: 'object',
          properties: {
            code: {
              type: 'string',
              description: 'The code to review',
            },
            contentType: {
              type: 'string',
              description: 'The type of content being reviewed',
            },
            language: {
              type: 'string',
              description: 'The programming language of the code',
            },
          },
          required: ['code', 'contentType'],
        },
        func: async ({ code, contentType, language }) => {
          const observation = Observation.createCodeObservation('user', code, language || 'text');
          observation.data.context = { type: contentType };
          const action = await editorAgent.process(observation);
          return JSON.stringify(action.data);
        },
      }),
    ];
    
    // Create the agent
    const prompt = ChatPromptTemplate.fromMessages([
      ['system', this.systemMessage],
      ['human', '{input}'],
    ]);
    
    const agent = createOpenAIFunctionsAgent({
      llm: model,
      tools,
      prompt,
    });
    
    this.agentExecutor = AgentExecutor.fromAgentAndTools({
      agent,
      tools,
      memory,
      verbose: true,
    });
    
    return this.agentExecutor;
  }
  
  /**
   * Process a document and create a vector store
   */
  private async processDocument(fileContent: string, fileName: string, provider: string, apiKey: string): Promise<string> {
    try {
      let loader;
      
      if (fileName.endsWith('.pdf')) {
        // For PDF files
        const uint8Array = Buffer.from(fileContent, 'base64');
        loader = new PDFLoader(uint8Array);
      } else {
        // For text files
        loader = new TextLoader(fileContent);
      }
      
      const docs = await loader.load();
      
      // Split the documents into chunks
      const textSplitter = new RecursiveCharacterTextSplitter({
        chunkSize: 1000,
        chunkOverlap: 200,
      });
      const splitDocs = await textSplitter.splitDocuments(docs);
      
      // Create embeddings based on the provider
      let embeddings;
      if (provider === 'openai') {
        embeddings = new OpenAIEmbeddings({ openAIApiKey: apiKey });
      } else {
        // Fallback to a free model for embeddings
        embeddings = new HuggingFaceInferenceEmbeddings();
      }
      
      // Create a vector store from the documents
      const vectorStore = await MemoryVectorStore.fromDocuments(splitDocs, embeddings);
      
      // Store the vector store with a unique ID
      const storeId = `${Date.now()}-${fileName}`;
      this.vectorStores.set(storeId, vectorStore);
      
      return storeId;
    } catch (error) {
      console.error('Error processing document:', error);
      throw error;
    }
  }
  
  /**
   * Process an observation and generate an action
   * @param observation The observation to process
   * @returns The action to take
   */
  async process(observation: Observation): Promise<Action> {
    try {
      // Extract message and context from observation
      let message = '';
      let context: any = {};
      
      if (observation.data.type === 'user_message') {
        message = observation.data.message;
        context = observation.data.context || {};
      } else if (observation.data.type === 'text') {
        message = observation.data.content;
        context = observation.data.context || {};
      } else {
        // For other observation types, convert to string
        message = JSON.stringify(observation.data);
      }
      
      // Get API settings from context or use defaults
      const provider = context?.apiSettings?.provider || process.env.DEFAULT_AI_PROVIDER || 'gemini';
      const apiKey = context?.apiSettings?.apiKey || process.env.GEMINI_API_KEY || process.env.OPENAI_API_KEY;
      
      if (!apiKey) {
        return Action.createErrorAction(this.id, 'No API key provided. Please configure your API key in settings.');
      }
      
      const responses: any[] = [];
      
      // Process any files in the context
      if (context?.files && context.files.length > 0) {
        for (const file of context.files) {
          try {
            const storeId = await this.processDocument(file.content, file.name, provider, apiKey);
            responses.push({
              type: 'agent',
              content: `Processed file: ${file.name}`,
              metadata: { fileId: storeId }
            });
          } catch (error) {
            console.error(`Error processing file ${file.name}:`, error);
            responses.push({
              type: 'error',
              content: `Failed to process file: ${file.name}`
            });
          }
        }
      }
      
      // Initialize the LangChain agent
      const agentExecutor = await this.initializeAgentExecutor(provider, apiKey);
      
      // Execute the agent
      const result = await agentExecutor.invoke({
        input: message,
        context: JSON.stringify(context)
      });
      
      // Parse the agent's output
      let agentResponses: any[] = [];
      
      try {
        // Try to parse the output as JSON
        const parsedOutput = JSON.parse(result.output);
        
        if (Array.isArray(parsedOutput)) {
          agentResponses = parsedOutput;
        } else {
          // If it's a single response object
          agentResponses = [parsedOutput];
        }
      } catch (error) {
        // If parsing fails, treat it as a text response
        agentResponses = [{
          type: 'text',
          content: result.output
        }];
      }
      
      // Add the agent responses to the final responses
      responses.push(...agentResponses);
      
      // If no responses were generated, add a fallback response
      if (responses.length === 0) {
        // Fallback to the traditional approach
        const thinkingAgent = this.agentRegistry.getAgent('thinking-agent');
        if (thinkingAgent) {
          const thinkingObservation = Observation.createUserMessageObservation('user', message);
          thinkingObservation.data.context = context;
          const thinkingAction = await thinkingAgent.process(thinkingObservation);
          
          responses.push(thinkingAction.data);
          
          if (thinkingAction.data.type !== 'error' && thinkingAction.data.type === 'plan') {
            // Extract the plan from the thinking agent's response
            const plan = thinkingAction.data.content;
            
            // Execute each step in the plan
            for (const step of plan.steps) {
              let stepAction: Action;
              
              switch (step.agent) {
                case 'DeveloperAgent': {
                  const developerAgent = this.agentRegistry.getAgent('developer-agent');
                  if (developerAgent) {
                    const devObservation = Observation.createUserMessageObservation('user', message);
                    devObservation.data.context = step.params;
                    stepAction = await developerAgent.process(devObservation);
                  } else {
                    stepAction = Action.createErrorAction(this.id, 'DeveloperAgent not found');
                  }
                  break;
                }
                case 'EditorAgent': {
                  // Find the most recent code response to edit
                  const codeResponses = responses.filter(r => r.type === 'code');
                  if (codeResponses.length > 0 && this.agentRegistry.getAgent('editor-agent')) {
                    const latestCode = codeResponses[codeResponses.length - 1];
                    const editorAgent = this.agentRegistry.getAgent('editor-agent');
                    const editorObservation = Observation.createCodeObservation(
                      'user',
                      latestCode.content,
                      latestCode.language || 'text'
                    );
                    editorObservation.data.context = { 
                      type: 'code',
                      language: latestCode.language
                    };
                    stepAction = await editorAgent.process(editorObservation);
                  } else {
                    stepAction = Action.createErrorAction(this.id, 'No code found to edit or EditorAgent not found');
                  }
                  break;
                }
                default:
                  stepAction = Action.createErrorAction(this.id, `Unknown agent: ${step.agent}`);
              }
              
              responses.push(stepAction.data);
              
              if (stepAction.data.type === 'error') {
                break;
              }
            }
          }
        } else {
          responses.push({
            type: 'error',
            content: 'ThinkingAgent not found and no responses generated'
          });
        }
      }
      
      // Create a composite action with all responses
      return new Action(this.id, {
        type: 'composite',
        responses: responses
      });
    } catch (error) {
      console.error('Error in OrchestratorAgent:', error);
      return Action.createErrorAction(this.id, 'Failed to orchestrate agents');
    }
  }
  
  /**
   * Legacy method for backward compatibility
   * @param message The message to process
   * @param context Additional context
   * @returns The final response after processing through all necessary agents
   * @deprecated Use process(observation) instead
   */
  async processMessage(message: string, context?: any): Promise<AgentResponse[]> {
    // Create an observation from the message
    const observation = Observation.createUserMessageObservation('user', message);
    
    // Add context to the observation if provided
    if (context) {
      observation.data.context = context;
    }
    
    // Process the observation
    const action = await this.process(observation);
    
    // Convert the composite action to legacy responses
    if (action.data.type === 'composite' && Array.isArray(action.data.responses)) {
      return action.data.responses.map((response: any) => {
        if (response.type === 'text') {
          return {
            type: 'text',
            content: response.content
          };
        } else if (response.type === 'code') {
          return {
            type: 'code',
            content: response.content,
            metadata: {
              language: response.language,
              filePath: response.filePath
            }
          };
        } else if (response.type === 'error') {
          return {
            type: 'error',
            content: response.content
          };
        } else if (response.type === 'plan') {
          return {
            type: 'plan',
            content: response.content
          };
        } else if (response.type === 'agent') {
          return {
            type: 'agent',
            content: response.content,
            metadata: response.metadata
          };
        } else {
          return {
            type: 'text',
            content: JSON.stringify(response)
          };
        }
      });
    } else {
      // If it's not a composite action, convert it to a single response
      return [this.actionToLegacyResponse(action)];
    }
  }
}

File: backend-fastapi/app/services/orchestrator_agent.py
"""
Orchestrator Agent for the NeuroNest AI system.
This agent coordinates other agents and tools to complete tasks.
"""

import os
import logging
import json
from typing import Dict, List, Any, Optional, Union
import google.generativeai as genai
from openai import OpenAI
from anthropic import Anthropic
from langchain.agents import AgentExecutor, create_openai_tools_agent, create_react_agent
from langchain.memory import ConversationBufferMemory
from langchain.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain.schema import SystemMessage, HumanMessage, AIMessage
from langchain_openai import ChatOpenAI
from langchain_google_genai import ChatGoogleGenerativeAI
from langchain_anthropic import ChatAnthropic

from app.core.config import settings
from app.models.agent import AgentThought
from app.services.base_agent import BaseAgent, EventStream, Action, Observation
from app.services.agent_tools import get_agent_tools
from app.services.file_service import extract_text_from_file

logger = logging.getLogger(__name__)

class OrchestratorAgent(BaseAgent):
    """
    Orchestrator Agent that coordinates other agents and tools
    """
    
    def __init__(self, event_stream: EventStream, llm_provider: Any):
        """
        Initialize a new OrchestratorAgent
        
        Args:
            event_stream: The event stream for the agent
            llm_provider: The LLM provider for the agent
        """
        super().__init__(event_stream, llm_provider)
        self.name = "OrchestratorAgent"
        self.description = "Coordinates other agents and tools to complete tasks"
        self.capabilities = [
            "Process natural language requests",
            "Generate code in multiple languages",
            "Create and manage projects",
            "Execute code in a secure environment",
            "Process and analyze files"
        ]
        self.tools = get_agent_tools()
    
    async def initialize(self, config: Dict[str, Any]) -> None:
        """
        Initialize the agent
        
        Args:
            config: Configuration for the agent
        """
        await super().initialize(config)
        
        # Set default system message if not provided
        if not self.system_message:
            self.system_message = """You are NeuroNest AI, an advanced AI assistant that helps users with various tasks.
You have access to tools that can help you complete tasks such as reading and writing files,
executing code, creating projects, and more. Always use the appropriate tool when needed.
If you don't know how to do something or if a tool is not available, be honest and say so.

When generating code, make sure it's well-structured, follows best practices, and is properly documented.
When creating projects, organize files in a logical structure and include necessary configuration files.
"""
    
    def _get_llm(self, provider: str, api_key: str, temperature: float = 0.7) -> Union[ChatOpenAI, ChatGoogleGenerativeAI, ChatAnthropic]:
        """
        Get the appropriate LLM based on the provider
        
        Args:
            provider: The provider name
            api_key: The API key for the provider
            temperature: The temperature for the LLM
            
        Returns:
            The LLM instance
        """
        if provider == "openai":
            return ChatOpenAI(
                model="gpt-4-turbo",
                temperature=temperature,
                api_key=api_key,
                streaming=True
            )
        elif provider == "gemini":
            return ChatGoogleGenerativeAI(
                model="gemini-pro",
                temperature=temperature,
                google_api_key=api_key,
                convert_system_message_to_human=True
            )
        elif provider == "anthropic":
            return ChatAnthropic(
                model="claude-3-opus-20240229",
                temperature=temperature,
                anthropic_api_key=api_key
            )
        else:
            raise ValueError(f"Unsupported provider: {provider}")
    
    def _create_agent_executor(
        self, 
        llm: Union[ChatOpenAI, ChatGoogleGenerativeAI, ChatAnthropic],
        tools: List[Any],
        memory: Optional[ConversationBufferMemory] = None,
        system_message: str = None
    ) -> AgentExecutor:
        """
        Create an agent executor with the appropriate tools and memory
        
        Args:
            llm: The LLM to use
            tools: The tools to use
            memory: The memory to use
            system_message: The system message to use
            
        Returns:
            The agent executor
        """
        if system_message is None:
            system_message = self.system_message
        
        prompt = ChatPromptTemplate.from_messages([
            ("system", system_message),
            MessagesPlaceholder(variable_name="chat_history"),
            ("human", "{input}"),
            MessagesPlaceholder(variable_name="agent_scratchpad")
        ])
        
        if isinstance(llm, ChatOpenAI):
            agent = create_openai_tools_agent(llm, tools, prompt)
        else:
            agent = create_react_agent(llm, tools, prompt)
            
        return AgentExecutor(
            agent=agent,
            tools=tools,
            memory=memory,
            verbose=True,
            return_intermediate_steps=True
        )
    
    async def process(self, observation: Observation) -> Action:
        """
        Process an observation and generate an action
        
        Args:
            observation: The observation to process
            
        Returns:
            The action to take
        """
        try:
            # Extract message from observation
            message = ""
            context = {}
            files = []
            user_id = None
            api_keys = {}
            
            if observation.data["type"] == "user_message":
                message = observation.data["message"]
                context = observation.data.get("context", {})
                files = observation.data.get("files", [])
                user_id = observation.data.get("user_id")
                api_keys = observation.data.get("api_keys", {})
            elif observation.data["type"] == "text":
                message = observation.data["content"]
                context = observation.data.get("context", {})
            else:
                # For other observation types, convert to string
                message = json.dumps(observation.data)
            
            # Default provider
            provider = settings.DEFAULT_AI_PROVIDER
            
            # Override with context if provided
            if context and "provider" in context:
                provider = context["provider"]
                
            # Get API key
            api_key = None
            if api_keys and provider in api_keys:
                api_key = api_keys[provider]
            elif provider == "gemini" and settings.GEMINI_API_KEY:
                api_key = settings.GEMINI_API_KEY
            elif provider == "openai" and settings.OPENAI_API_KEY:
                api_key = settings.OPENAI_API_KEY
            elif provider == "anthropic" and settings.ANTHROPIC_API_KEY:
                api_key = settings.ANTHROPIC_API_KEY
                
            if not api_key:
                return Action.create_error_action(
                    self.id,
                    f"No API key found for provider: {provider}. Please provide an API key in your settings or request."
                )
                
            # Process files if provided
            file_contents = []
            file_ids = []
            if files:
                for file_path in files:
                    try:
                        text = await extract_text_from_file(file_path)
                        if text:
                            file_contents.append(f"Content of file {os.path.basename(file_path)}:\n{text}")
                            file_ids.append(os.path.basename(file_path))
                    except Exception as e:
                        logger.error(f"Error extracting text from file: {e}")
                        return Action.create_error_action(
                            self.id,
                            f"Error processing file: {str(e)}"
                        )
            
            # Add file contents to message if available
            if file_contents:
                message = message + "\n\n" + "\n\n".join(file_contents)
                
            # Convert history to LangChain format
            chat_history = []
            history = context.get("history", [])
            if history:
                for msg in history:
                    if msg["role"] == "user":
                        chat_history.append(HumanMessage(content=msg["content"]))
                    elif msg["role"] == "assistant":
                        chat_history.append(AIMessage(content=msg["content"]))
                        
            # Create memory
            memory = ConversationBufferMemory(
                memory_key="chat_history",
                return_messages=True,
                input_key="input"
            )
            
            # Add history to memory
            if chat_history:
                memory.chat_memory.messages.extend(chat_history)
                
            # Create LLM
            try:
                llm = self._get_llm(provider, api_key)
                
                # Create agent executor with tools
                agent_executor = self._create_agent_executor(
                    llm=llm,
                    tools=self.tools,
                    memory=memory
                )
                
                # Add file information to the context
                input_context = {}
                if context:
                    input_context.update(context)
                if file_ids:
                    input_context["file_ids"] = file_ids
                if user_id:
                    input_context["user_id"] = user_id
                    
                # Execute the agent
                result = await agent_executor.ainvoke(
                    {
                        "input": message,
                        "context": json.dumps(input_context)
                    }
                )
                
                # Extract thoughts from intermediate steps
                thoughts = []
                for step in result.get("intermediate_steps", []):
                    action = step[0]
                    action_input = action.tool_input if hasattr(action, "tool_input") else action.args
                    action_output = step[1]
                    
                    thoughts.append(
                        AgentThought(
                            thought=f"I need to use the {action.tool} tool",
                            action=action.tool,
                            action_input=str(action_input),
                            observation=str(action_output)
                        )
                    )
                
                # Create a text action with the result
                action = Action.create_text_action(self.id, result["output"])
                
                # Add thoughts to the action data
                action.data["thoughts"] = [thought.dict() for thought in thoughts]
                
                return action
                    
            # Process with the appropriate provider if agent execution fails
            except Exception as e:
                logger.error(f"Error executing agent: {e}")
                logger.info("Falling back to direct provider API")
                
                if provider == "gemini":
                    result = await self._process_with_gemini(message, history, api_key)
                    return Action.create_text_action(self.id, result["content"])
                elif provider == "openai":
                    result = await self._process_with_openai(message, history, api_key)
                    return Action.create_text_action(self.id, result["content"])
                else:
                    return Action.create_error_action(
                        self.id,
                        f"Unsupported provider: {provider}"
                    )
        except Exception as e:
            logger.error(f"Error processing observation: {e}")
            return Action.create_error_action(
                self.id,
                f"Error processing your request: {str(e)}"
            )
    
    async def _process_with_gemini(
        self, 
        message: str, 
        history: Optional[List[Dict[str, str]]] = None,
        api_key: str = None
    ) -> Dict[str, Any]:
        """
        Process a message using Google's Gemini API
        
        Args:
            message: The message to process
            history: The conversation history
            api_key: The API key for Gemini
            
        Returns:
            The response from Gemini
        """
        try:
            # Configure the Gemini API
            genai.configure(api_key=api_key)
            
            # Create a model instance
            model = genai.GenerativeModel(
                model_name="gemini-pro",
                generation_config={
                    "temperature": settings.AGENT_TEMPERATURE,
                    "max_output_tokens": settings.AGENT_MAX_TOKENS,
                    "top_p": 0.95,
                }
            )
            
            # Convert history to Gemini format if provided
            gemini_history = []
            if history:
                for msg in history:
                    role = "user" if msg["role"] == "user" else "model"
                    gemini_history.append({"role": role, "parts": [msg["content"]]})
            
            # Create a chat session
            chat = model.start_chat(history=gemini_history)
            
            # Send the message and get the response
            response = chat.send_message(message)
            
            return {
                "content": response.text,
                "type": "text"
            }
        except Exception as e:
            logger.error(f"Error with Gemini API: {e}")
            raise
    
    async def _process_with_openai(
        self, 
        message: str, 
        history: Optional[List[Dict[str, str]]] = None,
        api_key: str = None
    ) -> Dict[str, Any]:
        """
        Process a message using OpenAI's API
        
        Args:
            message: The message to process
            history: The conversation history
            api_key: The API key for OpenAI
            
        Returns:
            The response from OpenAI
        """
        try:
            # Initialize the OpenAI client
            client = OpenAI(api_key=api_key)
            
            # Convert history to OpenAI format if provided
            messages = []
            if history:
                for msg in history:
                    role = "user" if msg["role"] == "user" else "assistant"
                    messages.append({"role": role, "content": msg["content"]})
            
            # Add the current message
            messages.append({"role": "user", "content": message})
            
            # Send the request to OpenAI
            response = client.chat.completions.create(
                model="gpt-4",
                messages=messages,
                temperature=settings.AGENT_TEMPERATURE,
                max_tokens=settings.AGENT_MAX_TOKENS,
            )
            
            return {
                "content": response.choices[0].message.content,
                "type": "text"
            }
        except Exception as e:
            logger.error(f"Error with OpenAI API: {e}")
            raise